modules/streams-core/src/main/scala/com.snowplowanalytics.snowplow/sinks/ListOfList.scala

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (c) 2023-present Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Snowplow Community License Version 1.0,
+ * and you may not use this file except in compliance with the Snowplow Community License Version 1.0.
+ * You may obtain a copy of the Snowplow Community License Version 1.0 at https://docs.snowplow.io/community-license-1.0
+ */
+package com.snowplowanalytics.snowplow.sinks
+
+import fs2.Chunk
+import cats.{Eval, Foldable, Monad, Monoid}
+import scala.collection.compat._
+
+/**
+ * A data structure that is efficient for most Snowplow streaming apps
+ *
+ * This is implemented as a `List[List[A]]`. But the inner Lists are hidden from the developer, so
+ * to force us into only using efficient methods.
+ *
+ * A `ListOfList` has these features:
+ *
+ *   - **Fast prepend** when building bigger batches from smaller batches e.g.
+ *     `batchesOfEvents.prepend(anotherBatch)`.
+ *   - **Fast folding** e.g. Foldable[ListOfList].foldMap(batches)(event => ???)
+ *
+ * It is ideal for situations where:
+ *
+ *   - We don't care about order of events within a batch
+ *   - We want to minimize how often we copy data structures
+ *   - We don't need fast lookup by index
+ *   - We want to batch up small batches into large batches of events
+ *
+ * It is deliberately missing a few features, so to force us into efficient usage patterns:
+ *
+ *   - No `.size` or `.length` methods. In Snowplow apps we manage batch size by other means.
+ *   - No `.traverse`. Instead we can use:
+ *
+ * ```
+ * Foldable[ListOfList].traverseUnordered(listOfList)(a => IO { ??? })
+ * ```
+ */
+class ListOfList[+A](private val value: List[List[A]]) extends AnyVal {
+
+  def isEmpty: Boolean = value.forall(_.isEmpty)
+
+  /** Fast prepend a batch to the beginning of this ListOfList */
+  def prepend[B >: A](elems: List[B]): ListOfList[B] =
+    ListOfList.of(elems :: value)
+
+  /**
+   * Apply a transformation function `f` to every element in the ListOfList
+   *
+   * The resulting `ListOfList` does not have the same order as the input List. This is helpful in
+   * Snowplow apps where order of events within batches is not important.
+   */
+  def mapUnordered[B](f: A => B): ListOfList[B] =
```

Comment:
It is not preserved!

```
scala> val lol = ListOfList.of(List(List(1,2,3), List(4,5,6)))
val lol: com.snowplowanalytics.snowplow.sinks.ListOfList[Int] = com.snowplowanalytics.snowplow.sinks.ListOfList@3fdd4a80

scala> lol.copyToIndexedSeq
val res1: IndexedSeq[Int] = Vector(1, 2, 3, 4, 5, 6)

scala> lol.mapUnordered(identity).copyToIndexedSeq
val res2: IndexedSeq[Int] = Vector(6, 5, 4, 3, 2, 1)
```

Reply:
I get it now, it is because of `.asIterable.toList` that it is preserved here:
```
def nonEmpty3 = {
    val input    = ListOfList.of(List(List(1, 2, 3), List(4, 5, 6)))
    val expected = List(100, 200, 300, 400, 500, 600)
    val result   = input.mapUnordered(_ * 100).asIterable.toList

    result must containTheSameElementsAs(expected)
  }
```

Reply:
Oh I see the confusion, but that's just a specs2 thing.

In specs2, `containTheSameElementsAs` only tests that all elements are present in any order.  The documentation is [here](https://etorreborre.github.io/specs2/guide/SPECS2-5.0.7/org.specs2.guide.Matchers.html) and it contains this example:

> Finally, if you want to get the differences between 2 traversables:
> `Seq(2, 4, 1) must containTheSameElementsAs(Seq(1, 4, 2))`